### ISSUE TITLE
An issue with plotting a tree using machu.treeplot()

### ISSUE THREAD

**Issue by Zeroo11** [wxguillo/machuruku#2]
When using machu.treeplot() to plot the tree, I am unable to display the complete tree. And it has a warning.
<img width="1278" alt="image" src="https://github.com/wxguillo/machuruku/assets/72841838/16aa25a8-05da-4d18-9e11-a92c27049c54">
Warning message:
Removed 1 rows containing missing values (`geom_point_g_gtree()`). 

However, when using plot(), it works fine. Could you please explain the reason for this? I am looking forward to your response. Thank you very much!
<img width="831" alt="image" src="https://github.com/wxguillo/machuruku/assets/72841838/88accc45-3b8a-4706-aca3-f0c3a4f2c5d7">

**Comment by wxguillo** [comment]
Hi Zeroo, I am working on version 2 of Machuruku and happen to have rewritten this function already to hopefully be less janky. The full release isn't ready yet though so for now you'll have to copy and paste my code into your Rstudio and load the function yourself.
The only library it relies on is phytools, so you'll have to load that. You can use the x.u.lim and x.l.lim parameters to add space around the tree in case yours doesn't fit. I believe the issue you were having is that your tree is pretty big, and you probably needed to adjust with the x.lim parameter more. Anyway, try the new function and see if it works. 
```
machu.treeplot <- function(tree, timeslice=NULL, nodelabs=T, nodelabsize=0.5, col="red", x.u.lim=NULL, x.l.lim=NULL, timelabs=T, timelabsize=0.8, timelaboffset=-0.35){
  # save user's previous par settings
  pars <- par() 
  
  # plot single tree if a list of length=1 is given, or a simple phylo object
  if (length(tree) == 1 || class(tree) == "phylo" ){
    
    if (class(tree)=="list") tree <- tree[[1]] # if tree is a list with length=1, unlist it
    
    par(mar=c(3.1, 2.1, 1.1, 1.1)) # set plot margins
    
    th <- max(nodeHeights(tree)) # calculate tree height
    om <- 10^floor(log10(th)) # calculate order of magnitude of tree height
    xlims <- c(ceiling(th/om)*om, -om) # dynamically calculate a good x axis
    if (is.null(x.l.lim)==F) xlims[1] <- x.l.lim # override lower x-limit
    if (is.null(x.u.lim)==F) xlims[2] <- x.u.lim # override upper x-limit
    ylims <- c(0.5, 1.2*Ntip(tree)) # dynamically calculate a good y axis
    
    plot(NA, xlim=xlims, ylim=ylims, bty="n", axes=F, xlab="", ylab="") # set up plot
    axis(1, at=seq(0, xlims[1], by=om)) # draw dynamic timescale
    abline(v=timeslice, col=col) # draw timeslices, if provided
    plotTree(tree, direction="leftwards", xlim=xlims, ylim=ylims, mar=par()$mar, color="black", add=T)
    if (nodelabs==T) labelnodes(text=1:tree$Nnode, node=1:tree$Nnode+Ntip(tree), circle.exp=nodelabsize, interactive=F)
    if (timelabs==T) ape::nodelabels(text=ape::branching.times(tree), node=1:tree$Nnode+Ntip(tree), 
                                     frame="none", adj=timelaboffset, cex=timelabsize)
  }
  # plot multiple trees if a list greater than length=1 is given
  if (class(tree) == "list" && length(tree) > 1){
    
    par(mar=c(0, 2.1, 0, 1.1), mfrow=c(length(trees), 1), xpd=NA) # set pars, including a column of 1 plot/tree
    
    th <- max(sapply(trees, function(x) max(nodeHeights(x)))) # get maximum node height from all trees
    om <- 10^floor(log10(th)) # calculate order of magnitude of tree height
    xlims <- c(ceiling(th/om)*om, -om) # dynamically calculate a good x axis
    if (is.null(x.l.lim)==F) xlims[1] <- x.l.lim # override lower x-limit
    if (is.null(x.u.lim)==F) xlims[2] <- x.u.lim # override upper x-limit
    
    # loop through each tree
    for (i in 1:length(tree)){ 
      # if on the last tree, add a bottom margin to give room for the timescale
      if (i==length(tree)) par(mar=c(3.1, 2.1, 0, 1.1))
      
      ylims <- c(0.5, 1.2*Ntip(trees[[i]])) # dynamically calculate a good y axis for each tree
      plotTree(tree[[i]], direction="leftwards", xlim=xlims, ylim=ylims, mar=par()$mar, color="black")
      if (nodelabs==T) labelnodes(text=1:tree[[i]]$Nnode, node=1:tree[[i]]$Nnode+Ntip(tree[[i]]),
                                  circle.exp=nodelabsize/2, interactive=F)
      if (timelabs==T) ape::nodelabels(text=ape::branching.times(tree[[i]]), node=1:tree[[1]]$Nnode+Ntip(tree[[i]]), 
                                       frame="none", adj=timelaboffset/0.8, cex=timelabsize/0.9)
    }
    axis(1, at=seq(0, xlims[1], by=om)) # plot timescale
    abline(v=timeslice, col=col) # plot timeslices. goes across all plots because par(xpd=T)
  }
  # return user's previous par settings
  par(pars)
}
```

**Comment by Zeroo11** [comment]
Thank you very much for your reply! I will try it later. 

Also, I have 2 more questions.
1. An error occurred when running machu.occ.rarefy(): 
```
> occ.rarefied <- machu.occ.rarefy(in.pts = occ, rarefy.dist = 10)
Error in structure(.External(.C_dotTclObjv, objv), class = "tclObj") : 
     [tcl] invalid command name "toplevel".
```
Is it because I'm using RstudioServer, which doesn't support tcl? (https://support.posit.co/hc/en-us/articles/226302647-Is-the-tcl-tk-package-supported-within-RStudio-Server-)

2. So I'm going to try machu.occ.rarefy() on my own Mac (M2), but the Machuruku package fails to install (R 4.3.0), with the following error:
```
> packageVersion("ggtree")
[1] ‘3.8.0’
> devtools::install_github("wxguillo/machuruku")
Downloading GitHub repo wxguillo/machuruku@HEAD
Skipping 1 packages not available: ggtree
── R CMD build ──────────────────────────────────────────────────────────────────
✔  checking for file ‘/private/var/folders/np/_zty4pm130j3k7w6v_dln9100000gn/T/RtmpEMK7Ka/remotes2d9174cca262/wxguillo-machuruku-b238ff0/DESCRIPTION’ ...
─  preparing ‘machuruku’:
✔  checking DESCRIPTION meta-information ...
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
   Omitted ‘LazyData’ from DESCRIPTION
─  building ‘machuruku_1.8.3.tar.gz’
   
* installing *source* package ‘machuruku’ ...
** using staged installation
** R
** byte-compile and prepare package for lazy loading
Warning: replacing previous import ‘TeachingDemos::%<%’ by ‘ggtree::%<%’ when loading ‘machuruku’
Warning: replacing previous import ‘ape::rotate’ by ‘ggtree::rotate’ when loading ‘machuruku’
Warning: replacing previous import ‘ggtree::rotate’ by ‘raster::rotate’ when loading ‘machuruku’
Warning: replacing previous import ‘ape::zoom’ by ‘raster::zoom’ when loading ‘machuruku’
Warning: replacing previous import ‘ggtree::flip’ by ‘raster::flip’ when loading ‘machuruku’
Warning: replacing previous import ‘raster::select’ by ‘tidytree::select’ when loading ‘machuruku’
tcltk DLL is linked to '/opt/X11/lib/libX11.6.dylib'
Error: .onLoad failed in loadNamespace() for 'tcltk', details:
  call: fun(libname, pkgname)
  error: X11 library is missing: install XQuartz from www.xquartz.org
Execution halted
ERROR: lazy loading failed for package ‘machuruku’
* removing ‘/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library/machuruku’
Warning message:
In i.p(...) :
  installation of package ‘/var/folders/np/_zty4pm130j3k7w6v_dln9100000gn/T//RtmpEMK7Ka/file2d9126b90bd6/machuruku_1.8.3.tar.gz’ had non-zero exit status
```
The ggtree package is already installed. And I did not find the Machuruku package on CRAN, is the ggtree version too high? Is Machuruku still maintained?

Sorry to bother you again. When is Machuruku version 2 released? Looking forward to it! !

**Comment by wxguillo** [comment]
> Is it because I'm using RstudioServer, which doesn't support tcl?

I've never used RstudioServer, so I'm not sure, but that does sound plausible. In the meantime I'd suggest using the "thin" function from the spThin package as an alternative.

> The ggtree package is already installed. And I did not find the Machuruku package on CRAN, is the ggtree version too high? Is Machuruku still maintained?

It could be a version incompatibility issue. Did you try installing ggtree using the command I provide on the Machuruku github page? 

```
install.packages("BiocManager")
BiocManager::install("ggtree")
```
I'm attempting to wean Machuruku off ggtree for version 2.0 because ggtree is always the source of so many issues, and I only used it for plotting (that's why I switched to using phytools for machu.treeplot). 
Machuruku is not on CRAN, just Github for now, because I don't currently have the time, energy, or skill to do the bug checks and such that CRAN requires. Perhaps someday though. 
I'm hoping to put Machuruku 2 out very soon, but I don't want to give you a hard date in case I can't make it! Stay tuned.
